### PR TITLE
Refactor for testability

### PR DESCRIPTION
**Pull request by frederoni**
This PR adds a few location managers that can be set on the `RouteController`
- `NavigationLocationManager`: This is the base location manager that handles permission and background mode. All other location managers inherits from this location manager.
- `DefaultLocationManager`: This location manager resides in the UI part of the framework because it needs access to UIKit to determine battery state and set the desired accuracy accordingly.
- `ReplayLocationManager`: Replays a list of locations.
- `SimulatedLocationManager`: Replaces SimulatedRoute.

`RouteController` now also supports simple dead reckoning using any of the location managers. However, `isDeadReckoningEnabled` defaults to `false`.

`NavigationMapView` now receives location updates from the location manager associated with the `RouteController` through the new `RouteControllerDelegate.routeController(_:didUpdateLocations:)`.

- [x] Implement ReplayLocationManager
- [x] Replace SimulatedRoute with SimulatedLocationManager
- ~Refactor monitorStepProgress~
- ~Write tests using ReplayLocationManager~

**The gist of how to use these location managers:**

`SimulatedLocationManager` - Simulates location updates along a route

<details>

```swift
let locationManager = SimulatedLocationManager(route: route)
let navigationViewController = NavigationViewController(for: route, locationManager: locationManager)
```

</details>

<hr>

`ReplayLocationManager` - Replays a list of locations previously recorded to a file.
<details>

```swift
let locationManager = ReplayLocationManager()
locationManager.locations = Array<CLLocation>.locations(from: filePath)
let navigationViewController = NavigationViewController(for: route, locationManager: locationManager)
```

</details>

### CODE REVIEW

File: Examples/Swift/ViewController.swift [hunk 1]
```diff
@@ -52,6 +52,18 @@ class ViewController: UIViewController, MGLMapViewDelegate, NavigationViewContro
         getRoute()
     }
     
+    @IBAction func didTapReplay(_ sender: Any) {
```
Comment (1ec5) [inline Examples/Swift/ViewController.swift]: An action is usually named something more active, like `replay(_:)`.

File: MapboxCoreNavigation/LocationManagers/NavigationLocationManager.swift [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    
+    override public init() {
+        super.init()
+        
+        if CLLocationManager.authorizationStatus() == .notDetermined {
+            requestWhenInUseAuthorization()
+        }
+        
+        if Bundle.main.backgroundModeLocationSupported {
+            activityType = .automotiveNavigation
```
Comment (1ec5) [inline MapboxCoreNavigation/LocationManagers/NavigationLocationManager.swift]: Should the activity type always be `automotiveNavigation`, even when the user is getting walking directions?

File: MapboxCoreNavigation/RouteController.swift [hunk 3]
```diff
@@ -51,6 +51,18 @@ public protocol RouteControllerDelegate: class {
...
     optional func routeController(_ routeController: RouteController, didFailToRerouteWith error: Error)
+    
+    /**
+     Called when the route controller’s location manager receive a location update.
+     
+     These locations can be modified due to replay or simulation but they can
+     also derive from regular location updates from a `CLLocationManager`.
+     
+     - parameter routeController: The route controller that received the new locations.
+     - parameter locations: The locations that were received from on of the location managers.
```
Comment (1ec5) [inline MapboxCoreNavigation/RouteController.swift]: Typo: /on/one/

File: MapboxCoreNavigation/RouteController.swift [hunk 4]
```diff
@@ -142,11 +157,46 @@ open class RouteController: NSObject {
...
+        
+        self.locationManager(self.locationManager, didUpdateLocations: [interpolatedLocation])
+    }
+    
     public func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
         guard let location = locations.last else {
             return
         }
         
+        delegate?.routeController!(self, didUpdateLocations: [location])
```
Comment (1ec5) [inline MapboxCoreNavigation/RouteController.swift]: Should `routeController(_:didUpdateLocations:)` be optional if we require its existence here?

File: MapboxCoreNavigation/LocationManagers/NavigationLocationManager.swift (hunk 2 above)
Comment (frederoni) [inline MapboxCoreNavigation/LocationManagers/NavigationLocationManager.swift]: Extended RouteOptions to convert from `MBDirectionsProfileIdentifier` to `CLActivityType`.

File: MapboxCoreNavigation/RouteController.swift (hunk 4 above)
Comment (frederoni) [inline MapboxCoreNavigation/RouteController.swift]: Should be optional

File: MapboxNavigation/NavigationViewController.swift [hunk 5]
```diff
@@ -165,15 +166,6 @@ public class NavigationViewController: NavigationPulleyViewController, RouteMapV
     public var routeController: RouteController!
     
     /**
-     `simulate` provides simulated location updates along the given route.
-     */
-    public var simulatesLocationUpdates: Bool = false {
```
Comment (ericrwolfe) [inline MapboxNavigation/NavigationViewController.swift]: Should we keep this bool for simplicity in the Navigation UI and have the didSet function optionally set the SimulatedLocationManager?

File: MapboxNavigation/RouteMapViewController.swift [hunk 6]
```diff
@@ -83,42 +78,25 @@ class RouteMapViewController: UIViewController, PulleyPrimaryContentControllerDe
...
         mapView.setUserLocationVerticalAlignment(.bottom, animated: false)
         mapView.setUserTrackingMode(.followWithCourse, animated: false)
-
-        if simulatesLocationUpdates {
-            mapView.locationManager.stopUpdatingLocation()
-            mapView.locationManager.stopUpdatingHeading()
-        }
+        
+        mapView.locationManager.stopUpdatingLocation()
+        mapView.locationManager.stopUpdatingHeading()
```
Comment (ericrwolfe) [inline MapboxNavigation/RouteMapViewController.swift]: These should move to `viewWillAppear` to fix animation issues. More here: https://github.com/mapbox/mapbox-navigation-ios/issues/303

File: Examples/Swift/ViewController.swift [hunk 7]
```diff
@@ -143,11 +156,15 @@ class ViewController: UIViewController, MGLMapViewDelegate, NavigationViewContro
...
         //navigationViewController.voiceController?.identityPoolId = "<#Your AWS IdentityPoolId. Remove Argument if you do not want to use AWS Polly#>"
         
-        navigationViewController.simulatesLocationUpdates = simulatesLocationUpdates
         navigationViewController.routeController.snapsUserLocationAnnotationToRoute = true
+        navigationViewController.routeController.isDeadReckoningEnabled = true
         navigationViewController.voiceController?.volume = 0.5
         navigationViewController.navigationDelegate = self
         
+        if let locationManager = locationManager {
+            navigationViewController.routeController.locationManager = locationManager
```
Comment (frederoni) [inline Examples/Swift/ViewController.swift]: I wonder if Swift have something like Obj-C’s null_resettable.

File: MapboxNavigation/NavigationViewController.swift (hunk 5 above)
Comment (frederoni) [inline MapboxNavigation/NavigationViewController.swift]: We can't do the same for ReplayLocationManager, so for consistency, I'd prefer not to expose this shorthand.

File: Examples/Swift/ViewController.swift (hunk 7 above)
Comment (1ec5) [inline Examples/Swift/ViewController.swift]: Yes: `locationManager!` with a custom getter or setter that handles `nil`.

File: MapboxCoreNavigation/RouteController.swift [hunk 8]
```diff
@@ -307,9 +368,6 @@ extension RouteController: CLLocationManagerDelegate {
                 strongSelf.routeProgress = RouteProgress(route: route)
                 strongSelf.routeProgress.currentLegProgress.stepIndex = 0
                 strongSelf.delegate?.routeController?(strongSelf, didRerouteAlong: route)
-                NotificationCenter.default.post(name: RouteControllerDidReroute, object: self, userInfo: [
-                    MBRouteControllerNotificationRouteKey: location
```
Comment (frederoni) [inline MapboxCoreNavigation/RouteController.swift]: Moved this to routeProgress didSet and changed `RouteKey` to `LocationKey`

**Comment by frederoni** [comment]
`CLLocationManager` can crash if it's deallocated in the same runloop it's initialized so I had to change the API a bit. `RouteController` and `NavigationViewController` is now initialized with an optional `CLLocationManager`.

File: MapboxCoreNavigationTests/MapboxCoreNavigationTests.swift [hunk 9]
```diff
@@ -29,49 +30,86 @@ class MapboxCoreNavigationTests: XCTestCase {
...
+        let navigation = RouteController(along: route, directions: directions, locationManager: locationManager)
         
         self.expectation(forNotification: RouteControllerAlertLevelDidChange.rawValue, object: navigation) { (notification) -> Bool in
             XCTAssertEqual(notification.userInfo?.count, 2)
             
             let routeProgress = notification.userInfo![RouteControllerAlertLevelDidChangeNotificationRouteProgressKey] as? RouteProgress
             let userDistance = notification.userInfo![RouteControllerAlertLevelDidChangeNotificationDistanceToEndOfManeuverKey] as! CLLocationDistance
             
-            return routeProgress?.currentLegProgress.alertUserLevel == .low && routeProgress?.currentLegProgress.stepIndex == 2 && round(userDistance) == 1758
+            return routeProgress?.currentLegProgress.alertUserLevel == .low && routeProgress?.currentLegProgress.stepIndex == 2 && round(userDistance) == 1786
```
Comment (frederoni) [inline MapboxCoreNavigationTests/MapboxCoreNavigationTests.swift]: @1ec5 @bsudekum Any idea how `userDistance` could've changed?

Comment (1ec5) [inline MapboxCoreNavigationTests/MapboxCoreNavigationTests.swift]: Not sure. Do you think it could be a race condition?

Comment (frederoni) [inline MapboxCoreNavigationTests/MapboxCoreNavigationTests.swift]: I'm getting 1786 on master (2c6ec67258dfc1902951845635465276ce4af261) 😕

Comment (frederoni) [inline MapboxCoreNavigationTests/MapboxCoreNavigationTests.swift]: screenshot 2017-06-26 22 22 43
That test was disabled in the scheme. 😄

**Pull request merged**